Userland/Libraries/LibSoftGPU/Clipper.cpp

```diff
@@ -17,40 +18,58 @@ namespace SoftGPU {
 {
-    if constexpr (plane == Clipper::ClipPlane::LEFT)
+    if constexpr (plane == Clipper::ClipPlane::left)
         return vertex.x() >= -vertex.w();
-    else if constexpr (plane == Clipper::ClipPlane::RIGHT)
+    else if constexpr (plane == Clipper::ClipPlane::right)
         return vertex.x() <= vertex.w();
-    else if constexpr (plane == Clipper::ClipPlane::TOP)
+    else if constexpr (plane == Clipper::ClipPlane::top)
         return vertex.y() <= vertex.w();
-    else if constexpr (plane == Clipper::ClipPlane::BOTTOM)
+    else if constexpr (plane == Clipper::ClipPlane::bottom)
         return vertex.y() >= -vertex.w();
-    else if constexpr (plane == Clipper::ClipPlane::NEAR)
+    else if constexpr (plane == Clipper::ClipPlane::near)
         return vertex.z() >= -vertex.w();
-    else if constexpr (plane == Clipper::ClipPlane::FAR)
+    else if constexpr (plane == Clipper::ClipPlane::far)
         return vertex.z() <= vertex.w();
     return false;
 }
 
+static bool point_within_user_plane(FloatVector4 const& vertex, FloatVector4 const& user_plane)
+{
+    return vertex.dot(user_plane) >= 0;
+}
+
 template<Clipper::ClipPlane plane>
-static constexpr GPU::Vertex clip_intersection_point(GPU::Vertex const& p1, GPU::Vertex const& p2)
+bool point_within_plane(GPU::Vertex const& vertex, FloatVector4 const& user_plane)
 {
-    constexpr FloatVector4 clip_plane_normals[] = {
-        { 1, 0, 0, 1 },  // Left Plane
-        { -1, 0, 0, 1 }, // Right Plane
-        { 0, -1, 0, 1 }, // Top Plane
-        { 0, 1, 0, 1 },  // Bottom plane
-        { 0, 0, 1, 1 },  // Near Plane
-        { 0, 0, -1, 1 }  // Far Plane
-    };
-    constexpr auto clip_plane_normal = clip_plane_normals[to_underlying(plane)];
+    if constexpr (plane == Clipper::ClipPlane::user)
+        return point_within_user_plane(vertex.eye_coordinates, user_plane);
+    else
+        return point_within_clip_plane<plane>(vertex.clip_coordinates);
+}
 
-    // See https://www.microsoft.com/en-us/research/wp-content/uploads/1978/01/p245-blinn.pdf
-    // "Clipping Using Homogeneous Coordinates" Blinn/Newell, 1978
-    // Clip plane normals have W=1 so the vertices' W coordinates are included in x1 and x2.
+template<Clipper::ClipPlane plane>
+static GPU::Vertex clip_intersection_point(GPU::Vertex const& p1, GPU::Vertex const& p2, FloatVector4 const& plane_normal)
+{
+    float x1;
+    float x2;
+    switch (plane) {
+    case Clipper::ClipPlane::left:
+    case Clipper::ClipPlane::right:
+    case Clipper::ClipPlane::top:
+    case Clipper::ClipPlane::bottom:
+    case Clipper::ClipPlane::near:
+    case Clipper::ClipPlane::far:
+        x1 = plane_normal.dot(p1.clip_coordinates);
+        x2 = plane_normal.dot(p2.clip_coordinates);
+        break;
+    case Clipper::ClipPlane::user:
+        x1 = plane_normal.dot(p1.eye_coordinates);
+        x2 = plane_normal.dot(p2.eye_coordinates);
+        break;
+    default:
+        VERIFY_NOT_REACHED();
+    }
```

Comment:
Since the calculations are the same and you're just selecting different coordinates, maybe this is a bit neater?
```cpp
auto p1_coordinates = (plane == Clipper::ClipPlane::User) ? p1.eye_coordinates : p1.clip_coordinates;
auto p2_coordinates = (plane == Clipper::ClipPlane::User) ? p2.eye_coordinates : p2.clip_coordinates;
auto x1 = plane_normal.dot(p1_coordinates);
auto x2 = plane_normal.dot(p2_coordinates);
```

Reply:
I agree! I've updated to use this version in the latest commit.